backend/src/main/java/ai/verta/modeldb/experimentRun/subtypes/VersionInputHandler.java

```diff
@@ -0,0 +1,245 @@
+import ai.verta.modeldb.Location;
+import ai.verta.modeldb.ModelDBConstants;
+import ai.verta.modeldb.VersioningEntry;
+import ai.verta.modeldb.common.exceptions.InternalErrorException;
+import ai.verta.modeldb.common.exceptions.ModelDBException;
+import ai.verta.modeldb.common.futures.FutureJdbi;
+import ai.verta.modeldb.common.futures.InternalFuture;
+import ai.verta.modeldb.entities.versioning.CommitEntity;
+import ai.verta.modeldb.utils.ModelDBHibernateUtil;
+import ai.verta.modeldb.utils.ModelDBUtils;
+import ai.verta.modeldb.versioning.Blob;
+import ai.verta.modeldb.versioning.BlobDAO;
+import ai.verta.modeldb.versioning.BlobExpanded;
+import ai.verta.modeldb.versioning.CommitDAO;
+import ai.verta.modeldb.versioning.RepositoryDAO;
+import ai.verta.modeldb.versioning.RepositoryIdentification;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.hibernate.Session;
+
+public class VersionInputHandler {
+  private static Logger LOGGER = LogManager.getLogger(VersionInputHandler.class);
+
+  private final Executor executor;
+  private final FutureJdbi jdbi;
+  private final String entity_type;
+  private final String entityIdReferenceColumn;
+  private final RepositoryDAO repositoryDAO;
+  private final CommitDAO commitDAO;
+  private final BlobDAO blobDAO;
+  private static final ModelDBHibernateUtil modelDBHibernateUtil =
+      ModelDBHibernateUtil.getInstance();
+
+  public VersionInputHandler(Executor executor, FutureJdbi jdbi, String entityName, DAOSet daoSet) {
+    this.executor = executor;
+    this.jdbi = jdbi;
+    this.entity_type = entityName;
+
+    this.repositoryDAO = daoSet.repositoryDAO;
+    this.commitDAO = daoSet.commitDAO;
+    this.blobDAO = daoSet.blobDAO;
+
+    switch (entityName) {
+      case "ExperimentRunEntity":
+        this.entityIdReferenceColumn = "experiment_run_id";
+        break;
+      default:
+        throw new InternalErrorException("Invalid entity name: " + entityName);
+    }
+  }
+
+  public InternalFuture<ExperimentRun> validateAndInsertVersionedInputs(
+      ExperimentRun experimentRun) {
+    if (experimentRun.hasVersionedInputs()) {
+      InternalFuture<Map<String, Map.Entry<BlobExpanded, String>>> versionedInputFutureTask =
+          validateVersioningEntity(experimentRun.getVersionedInputs());
+      return versionedInputFutureTask
+          .thenCompose(
+              locationBlobWithHashMap ->
+                  insetVersioningInput(
+                      experimentRun.getVersionedInputs(),
+                      locationBlobWithHashMap,
+                      experimentRun.getId()),
+              executor)
+          .thenCompose(unused -> InternalFuture.completedInternalFuture(experimentRun), executor);
+    } else {
+      return InternalFuture.completedInternalFuture(experimentRun);
```

Comment:
Don't return the argument you just received.

Reply:
Fixed